permuta/perm_sets/permset.py

```diff
@@ -1,92 +1,168 @@
-            return PermSetStatic(descriptor)
+import multiprocessing
+from itertools import islice
+from typing import ClassVar, Dict, Iterable, List, NamedTuple, Optional, Union
 
-    @classmethod
-    def _dispatch_described(cls, descriptor):
-        # Loop through all the described superclasses; e.g. Avoiding
-        for described_superclass in PermSetDescribed.__subclasses__():
-            # Check if descriptor's class is of that of the superclasses'
-            # descriptor
-            if isinstance(descriptor, described_superclass.DESCRIPTOR_CLASS):
-                # The correct superclass has been found!
-                # Try to find subclass specifically for this descriptor
-                described_class = cls._find_described_class(
-                    descriptor, PermSetDescribed
-                )
-                if described_class is None:
-                    return described_superclass.DEFAULT_CLASS(descriptor)
-                else:
-                    return described_class(descriptor)
-        # TODO: Something else?
-        raise RuntimeError(
-            "PermSet for descriptor {} not found".format(repr(descriptor))
-        )
+from ..patterns import MeshPatt, Perm
+from ..permutils import is_finite, is_insertion_encodable, is_polynomial
+from .basis import Basis, MeshBasis
 
-    @classmethod
-    def _find_described_class(cls, descriptor, current_class):
-        # TODO: Use metaclasses to track subclasses, rather than this mess
-        if descriptor == current_class.DESCRIPTOR:
-            return current_class
-        else:
-            for subclass in current_class.__subclasses__():
-                described_class = cls._find_described_class(descriptor, subclass)
-                if described_class is not None:
-                    return described_class
-            return None
 
-    @classmethod
-    def avoiding(cls, basis):
-        BasisCls = detect_basis_cls(basis)
-        return cls(BasisCls(basis))
+class AvBase(NamedTuple):
+    """A base class for Av to define instance variables without having to use
+    __init__ in Av.
+    """
 
+    basis: Union[Basis, MeshBasis]
+    cache: List[Dict[Perm, Optional[List[int]]]]
 
-class AvoidanceClassMetaclass(type):
-    def __instancecheck__(self, instance):
-        return isinstance(instance, Avoiding)
 
-    def __subclasscheck__(self, subclass):
-        return issubclass(subclass, Avoiding)
+class Av(AvBase):
+    """A permutation class defined by its minimal basis."""
 
+    _CLASS_CACHE: ClassVar[Dict[Union[Basis, MeshBasis], "Av"]] = {}
+    _CACHE_LOCK = multiprocessing.Lock()
 
-class Av(object, metaclass=AvoidanceClassMetaclass):
-    def __new__(cls, basis=None):
-        if basis is None or len(basis) == 0:
-            return PermSetAll()
-        return PermSet.avoiding(basis)
+    def __new__(cls, basis: Union[Basis, MeshBasis]) -> "Av":
+        assert len(basis) > 0 and basis != Basis(Perm())
```

Comment:
I'd rather raise a ValueError here instead of asserting. I think we want the user to get some info and also about what is going on